neonize issue at startup

I can't start the server - this error pops up at startup and the server shuts down

INFO:     Started server process [28056]
INFO:     Waiting for application startup.
[14:47:05] INFO     📡 Subscribed to websocket outbound
           INFO     🔌 WebSocket Adapter subscribed to System Events
           INFO     Agent Loop started
           INFO     📡 Subscribed to whatsapp outbound
           INFO     🤖 Agent Loop started (Backend: open_interpreter)
panic: failed to dial whatsapp web websocket: failed to WebSocket dial: failed to send handshake request: Get "https://web.whatsapp.com/ws/chat": dial tcp 127.0.0.1:443: connectex: No connection could be made because the target machine actively refused it.

goroutine 17 [running, locked to thread]:
main.Neonize(0x249672455e0, 0x249672205f0, 0x7ffc8d7093c8, 0x0, 0x7ffc8d7093c8, 0x24967290fc0, 0x24967290f88, 0x24967290f50, 0x24967290f18, 0x24961dd4100, ...)
        /home/runner/work/neonize/neonize/goneonize/main.go:1039 +0xb9d

Update: when the VPN was turned on, everything worked.

--- a/src/pocketpaw/bus/adapters/neonize_adapter.py
+++ b/src/pocketpaw/bus/adapters/neonize_adapter.py
@@ -9,6 +9,7 @@ Created: 2026-02-06
 
 import asyncio
 import logging
+import socket
 import threading
 from pathlib import Path
 from typing import Any
@@ -46,6 +47,16 @@ def _ensure_neonize_loop_running() -> None:
         thread.start()
         _neonize_loop_started = True
         logger.debug("neonize event_global_loop started in background thread")
+
+
+def _tcp_probe(host: str, port: int, timeout: float) -> None:
+    """Blocking TCP connect probe — run in an executor."""
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.settimeout(timeout)
+    try:
+        sock.connect((host, port))
+    finally:
+        sock.close()
 
 
 class NeonizeAdapter(BaseChannelAdapter):
@@ -176,6 +187,12 @@ class NeonizeAdapter(BaseChannelAdapter):
         self._client = client
         self._jid2string = Jid2String
 
+        # Pre-flight: verify we can reach WhatsApp servers before letting
+        # neonize's Go runtime attempt the WebSocket dial.  A Go panic on
+        # connection failure kills the entire process, so we check first and
+        # raise a catchable Python exception instead.
+        await self._preflight_connectivity_check()
+
         # Schedule connect() on neonize's own event loop (not FastAPI's)
         from neonize.aioze.events import event_global_loop
 
@@ -186,6 +203,29 @@ class NeonizeAdapter(BaseChannelAdapter):
         # Give the client a moment to initialize and produce a QR code
         await asyncio.sleep(3)
         logger.info("WhatsApp (neonize) Adapter started — scan QR code to pair")
+
+    @staticmethod
+    async def _preflight_connectivity_check(
+        host: str = "web.whatsapp.com", port: int = 443, timeout: float = 5.0
+    ) -> None:
+        """Verify WhatsApp servers are reachable before connecting.
+
+        Neonize wraps whatsmeow (Go).  If the WebSocket dial to
+        web.whatsapp.com fails, the Go runtime panics and kills the entire
+        Python process.  A quick TCP probe lets us fail gracefully with a
+        normal exception instead.
+        """
+        loop = asyncio.get_running_loop()
+        try:
+            await asyncio.wait_for(loop.run_in_executor(
+                None, _tcp_probe, host, port, timeout
+            ), timeout=timeout + 1)
+        except Exception as exc:
+            raise ConnectionError(
+                f"Cannot reach {host}:{port} — WhatsApp (neonize) adapter "
+                f"requires internet access to web.whatsapp.com. "
+                f"Check your network or VPN settings."
+            ) from exc
 
     async def _on_stop(self) -> None:
         """Stop neonize client."""

--- a/tests/test_neonize_adapter.py
+++ b/tests/test_neonize_adapter.py
@@ -290,3 +290,32 @@ async def test_bus_integration(bus):
     adapter.send.assert_called_once_with(msg)
 
     await adapter.stop()
+
+
+async def test_preflight_check_passes_when_reachable():
+    """Preflight check succeeds when the host is reachable."""
+    with patch(
+        "pocketpaw.bus.adapters.neonize_adapter._tcp_probe", return_value=None
+    ):
+        # Should not raise
+        await NeonizeAdapter._preflight_connectivity_check()
+
+
+async def test_preflight_check_raises_when_unreachable():
+    """Preflight check raises ConnectionError when host is unreachable."""
+    with patch(
+        "pocketpaw.bus.adapters.neonize_adapter._tcp_probe",
+        side_effect=OSError("Connection refused"),
+    ):
+        with pytest.raises(ConnectionError, match="Cannot reach"):
+            await NeonizeAdapter._preflight_connectivity_check()
+
+
+async def test_preflight_error_message_mentions_vpn():
+    """Error message hints at VPN/network issues."""
+    with patch(
+        "pocketpaw.bus.adapters.neonize_adapter._tcp_probe",
+        side_effect=OSError("timed out"),
+    ):
+        with pytest.raises(ConnectionError, match="VPN"):
+            await NeonizeAdapter._preflight_connectivity_check()

